projecttacoma/bulk-export-server
# Add support for _typeFilter using search parameters

# Summary
Adds additional `_typeFilter` query support, now supporting a wide array of FHIR REST queries for each resource type using the projecttacoma fork of node-fhir-server-core and [Asymmetrik’s FHIR Query Builder](https://www.npmjs.com/package/@asymmetrik/fhir-qb).

## New behavior
The Bulk Export _typeFilter parameter, while still experimental, has the potential to greatly impact export performance. The _typeFilter parameter allows us to retrieve the tightest data set possible, as we can filter resource types by date fields, status, etc.

Right now, our server only supports code:in/type:in queries, and our _typeFilter query support is very specific to these kinds of queries. If we try to do another query, such as `MedicationRequest?status=active`, our server will just ignore this filter.

The updates in this PR allow the client to use _typeFilter values of the following formats:
•	Filter by search parameter for a given resource type (ex. `_typeFilter=MedicationRequest?status=active`)
•	Filter by multiple search parameters for a given resource type (ex. `_typeFilter=MedicationRequest?status=active&date=gt2018-07-01T00:00:00Z`)
•	Filter by conditional “or” search parameter queries for a given resource type (ex.  `_typeFilter=MedicationRequest?status=active,MedicationRequest?status=completed` to check for `MedicationRequest` resources with status “active” OR status “completed”)
•	Filter on multiple resource types via search parameters (ex. `_typeFilter=MedicationRequest? date=gt2018-07-01T00:00:00Z,Encounter?status=completed`

These queries can be used for system-level, group-level, and patient-level export.
## Code changes
**README**
- added a section on supported query parameters for the server

**Export.service.js**
- Wrapped all errors in `OperationOutcome` resources. The spec mentions in several places that the server should return an OperationOutcome resource in an error response when errors arise (ex. When an unsupported resource type is provided)
- Added a check on resource types that are included in the `_typeFilter` parameter

**ExportToNDJson.js**
Update `exportToNDJson` function to now do the following:
- Now acknowledges “search parameter” queries. These are `_typeFilter` queries that contain valid search parameters for the given resource type – for example, checking the `status` on `MedicationRequest` resources. These queries are indexed to grab the “property” and “property value” and pass them into the Asymmetrik FHIR Query Builder. For example, if the filter provided is `_typeFilter=MedicationRequest?status=completed&date=gt2018-07-01T00:00:00Z`, the function extracts the resource type (`MedicationRequest`) and extracts all subqueries (`status: “completed”` and `date`:  “gt2018-07-01T00:00:00Z”`).
- Continues to acknowledge the ValueSet-related `_typeFilter` queries that we currently support. An example query would be `_typeFilter=Encounter?type:in=http://some-value-set-url`. These queries are similarly parsed but are not passed through the query builder – instead, they get processed through the server to extract the codes from the ValueSet using existing server functions.

Updates `getDocuments` function to now do the following:
- For search parameter queries, runs an aggregation query to extract all the resources that match the queries. If ValueSet queries or patient filters also need to be applied, they are integrated into the aggregation pipeline.
- Miscellaneous changes to update `let` to `const` for many variables whose values never change, update variable names and function names to be more representative of their functionality, etc.

**valueSetHelper.js**
- Updates `getHierarchicalCodes` function to loop over an array of FHIR `ValueSet`s, as an array is passed into the function but is not appropriately accounted for within the function. This is a bug that was detected when testing the current functionality of the `ValueSet` filtering with `_typeFilter`

**Updated tests**
- export.server.test.js` tests now check for `OperationOutcome` responses
- `exportToNDJSON.test.js` changes
: (1) Create new Condition test fixture that contains two fields that can be filtered on , (2) Create new tests for `getDocuments` to check that it can filter based on multiple queries

# Testing guidance
First, run unit tests.
We will want to test a lot of combinations of `_typeFilter`, `_type`, and levels of export (system, group, and patient). Here’s an outline of steps that can be followed for testing:

1.	Upload resources to the server. This can be done with the `upload-bundles` script or `curl`. Make sure to upload groups of patients (the `upload-bundles` script does this automatically). The `post-txn-bundles` script can also be used to upload transaction bundles to the server, although it may need some tweaking to accommodate the group sizes you are uploading and your working directory.
2.	Run the server with `npm start`.
3.	Use Insomnia or a client like [the bulk data CLI Export Client](https://github.com/bulk-dqm/bulk-data-export-client) for sending requests to the server.

Requests to send to the server include, but are not limited to:
•	System-level export to ensure there are no errors
•	System-level export with `_type`
•	System-level export with `_typeFilter`
•	System-level export with both `_type` and `_typeFilter`

•	Group-level export to ensure there are no errors
•	Group-level export with `_type`
•	Group-level export with `_typeFilter`
•	Group-level export with both `_type` and `_typeFilter`

•	Patient-level export to ensure there are no errors
•	Patient-level export with `_type`
•	Patient -level export with `_typeFilter`
•	Patient -level export with both `_type` and `_typeFilter`

For the `_typeFilter` queries, test the following:
•	Query with “&” operator and check that the resources returned abide by all the conditions
•	Query with comma-delimited subqueries and check that the resources returned abide by at least one of the conditions
•	Test that the ValueSet filtering still works. To do this, you can refer to the data requirements of a given measure and extract a ValueSet url to include in the `_typeFilter` parameter. For example, you could do `_typeFilter=Encounter?type:in=<valueset-url>`

**Note**: Depending on the test environment that you are using, you may need to escape certain characters (like the `?` or `&` in the `_typeFilter` query) or manually URL encode them. Here are some cases I have come across when testing:
* From the terminal - you may need to escape `?` and `&` symbols (ex. `_typeFilter=MedicationRequest\?status=active`)
* From Insomnia - if you put the `_typeFilter` parameter and parameter value under the “Query” tab, you do not need to do any URL encoding - it gets picked up by Insomnia automatically. However, if you place the query within the request URL field (ex. Place `<url>/$export?_typeFilter=MedicationRequest?status=active` directly in the URL field), you will need to apply the URL encoding manually to the symbols, or else they will be interpreted to be separate parameters by Insomnia. (Ex. You may need to replace `&` with its URL encoding of `%26` in your request URL)
